rfd/0191-workload-id-config-ux.md

```diff
@@ -0,0 +1,868 @@
+perform benchmarks and performance improvements with growing magnitudes of 
+scale in mind.
+
+The flexible templating mechanism is intended to limit the number of Roles,
+Bots and WorkloadIdentitys used for large-scale deployment. For example,
+thousands of CI workflows can share the same Bot and WorkloadIdentity.
+Where possible, we can mitigate the need for growing numbers of WorkloadIdentity
+resources by adding more advanced templating and authorization rule
+functionality.
+
+We should bear in mind the following potential performance improvements:
+
+- Tailored cache with label-based indexing to improve the performance of 
+  resolving label matchers to specific WorkloadIdentity resources.
+- Edge cache in `tbot` that can be enabled to return recently-issued identities
+  without hitting the Auth Service. This will involve some elements of the
+  authorization logic being executed at the edge for the cache.
+- Decoupling the Workload Identity CA and issuance process from the Auth
+  Service.
+
+OpenTelemetry tracing should be included across the credential issuance process
+to highlight performance hot-spots.
+
+### Deprecation
+
+For now, the old RBAC and RPCs for Workload Identity will be left in place as
+they provide effectively an "administrative" way of generating workload 
+identity credentials.
+
+When the new mechanism has been available for 2 major versions, we should
+revisit the deprecation of the legacy RPCs and RBAC.
+
+## Alternatives
+
+### Role Templating and `tbot` SPIFFE ID templating
+
+One alternative to introducing a new resource is to introduce support for role
+templating of the SPIFFE fields.
+
+This would still require work to propagate the join metadata into the
+certificate as trait-like things for the purposes of RBAC evaluation.
+
+We'd also need to implement similar templating into the `tbot` configuration
+itself, we could likely reuse much of the same functionality here.
+
+Positives:
+
+1. "Simpler" to implement (but not significantly so)
+2. Keeps configuration within an existing Teleport primitive (roles) rather than
+  further spreading it out into a new resource.
+
+Negatives:
+
+1. TBot configuration still remains important in the SVID issuance process,
+  meaning you still need to manage configuration for a large fleet of `tbot`
+  instances.
+2. The configuration is split between Teleport resources and the `tbot` 
+  configuration. If these drift, SVIDs will fail to be issued.
+3. Teleport Roles are already pretty complex, and this is introducing further
+  complexity.
+
+## Security Considerations
+
+### Auditing
+
+The CRUD RPCs for the WorkloadIdentity resource will emit the standard audit 
+events:
+
+- `workload_identity.create`
+- `workload_identity.update`
+- `workload_identity.delete`
+
+The new RPC for issuing a credential based on a WorkloadIdentity will emit a
+new audit event `workload_identity.generate` which will contain the following
```

Comment:
Should this be emitted per workload identity credential - or - per call of the RPC ( think cases where a label selector has been provided ). I'm leaning towards per-workload-identity-credential as this yields the same result structure regardless of whether name selector label selector is in use.

My only other thought is whether we want to emit a 'failure' event - I don't think we can do this in cases where the label selectors have been used (since there will be a large number of failures to find one success in some cases) but we could do this for cases where a specific WorkloadIdentity has been named. This could help with diagnostics - but - I think would be weird to be inconsistent with the label selector mode.

Reply:
Per credential seems the natural choice with a bit of thought - I'll edit the RFD to reflect this.